repoman/git/repository.py

```diff
@@ -335,29 +335,33 @@ def push(self, orig, dest, rev=None, ref_name=None, force=False):
         """Inherited method
         :func:`~repoman.repository.Repository.push`
         """
-        if rev == None and ref_name == None:
+        all_tags_option = "--tags"
+        all_notes_refspec = "refs/notes/*:refs/notes/*"
+
+        if rev is None and ref_name is None:
             # Push everything
             refspec = "refs/*:refs/*"
-        elif rev == None:
+        elif rev is None:
             refspec = "%s:%s" % (ref_name, ref_name)
-        elif ref_name == None:
-            raise RepositoryError("When pushing, revision specified but not reference name")
+        elif ref_name is None:
+            raise RepositoryError(
+                "When pushing, revision specified but not reference name")
         else:
-            remote_refs = list(
-                self._git("ls-remote", "--tags", "--heads",
-                          dest, ref_name,
-                          _iter=True))
-            if len(remote_refs) == 0:
-                # Reference is not qualified
-                if self.tag_exists(ref_name):
-                    # We don't know what this ref is in remote, but here it is a tag
-                    ref_name = "refs/tags/%s" % ref_name
-                else:
-                    # In any other case, we assume it is a branch
-                    ref_name = "refs/heads/%s" % ref_name
+            if self.tag_exists(ref_name):
+                # We don't know what this ref is in remote, but here it is a tag
+                ref_name = "refs/tags/%s" % ref_name
+                all_tags_option = ""
```

Comment:
So, when the ref_name references a tag (it exists), `all_tags_option` is set to empty and therefore the command in 360 line (within "if") will not be executed as someone could think. 
But, when the reference is no a tag , then we use that variable and so "--tags" parameter.

I see it difficult to follow.